inflation/dataset/parse.py

```diff
@@ -0,0 +1,386 @@
+    product_brand: str
+    price: float
+    currency: str
+    in_stock: bool
+    sample_date: datetime.datetime
+
+
+class InflationDataset:
+    def __init__(self, data: list):
+        self.dataset = data
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def __iter__(self):
+        return InflationDatasetIterator(self.dataset)
+
+    def __getitem__(self, s):
+        if s < 0:
+            s = self.__len__() - s
+            if s < 0 or s >= self.__len__():
+                raise IndexError
+            else:
+                return self.dataset[s]
+        else:
+            start, stop, step = s.indices(self.__len__())
+            indxs = list(range(start, stop, step))
+            return [self.dataset[i] for i in indxs]
+
+    def save_dataset(self, output_file_name):
+        """
+        It saves dataset records by serializing.
+        """
+        full_path = SAVED_DATASETS_DIR_PATH / output_file_name
+
+        with open(full_path, "wb") as f:
+            cPickle.dump(self, f)
+
+    @classmethod
+    def read_dataset(cls, file_path):
+        with open(file_path, "rb") as input_file:
+            return cPickle.load(input_file)
+
+    def to_df(self):
+        return pd.DataFrame([record.__dict__ for record in self.dataset])
+
+
+class InflationDatasetIterator:
+    def __init__(self, inflation_dataset):
+        self.dataset = inflation_dataset
+        self._i = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self._i >= len(self.dataset):
+            raise StopIteration
+        else:
+            result = self.dataset[self._i]
+            self._i += 1
+            return result
+
+
+class ParserManager:
+    def __init__(self, parsers: dict):
+        self.parsers = parsers
+
+    def start_parsing_from_google_storage(
+        self,
+        google_storage_client,
+        data_file_path: str,
+        destination_file_path: str,
+    ):
+        """
+
+        Args:
+            data_file_path (str): The data file path in the Google Storage.
+            destination_file_path (str): The destination file path for saving the file (gzip
+            or json file format).
+
+        Returns:
+
+        """
+        google_storage_client.download(data_file_path, destination_file_path)
+        return self._start_parsing(destination_file_path)
+
+    def start_parsing_from_drive(self, data_file_path: str):
+        return self._start_parsing(data_file_path)
+
+    def _start_parsing(self, data_file_path: str):
+        records = []
+
+        with ParserManager.open(data_file_path, mode="rt") as file:
+            total_pages = 0
+            item_pages = 0
+            for line in file:
+                line = json.loads(line)
+                total_pages += 1
+
+                records.append(self.parsers[line["source"].lower()].parse(line))
+
+                item_pages += 1
+        logger.info(f"total_pages={total_pages}, item_pages={item_pages}")
```

Comment:
be mindful about your printing? ask yourself: this information should be enough for me to understand what's wrong immediately? Improve the logging: "Parsing is done: {item_pages}/{total_pages} parsed to {data_file_path}"